### PR TITLE
Added some date and time formatting options to scrolling text effect.

### PR DESCRIPTION
**Pull request by 5chubrakete**
Added/updated the following tags to use with the scrolling text effect.

 #hh Hours
 #mm Minutes
 #ss Seconds

 #DD Day
 #MM Month
 #MMM Month as Text
 #YYYY Year
 #YY Two digit Year

Changed #hhmm to lower case to not get confused with month.

This enables more flexible display of time and date on a bigger matrix. Also beeing able to display seconds in smaller font or lower brightness.

### CODE REVIEW

**Comment by blazoncek** [comment]
Why not just uppercase `text` (if it starts with #) and add "#YY" and "#YYYY"?

File: wled00/FX.cpp [hunk 1]
```diff
@@ -5960,10 +5960,16 @@ uint16_t mode_2Dscrollingtext(void) {
     else if (!strncmp_P(text,PSTR("#DDMM"),5)) sprintf_P(text, zero?PSTR("%02d.%02d")     :PSTR("%d.%d"),      day(localTime),   month(localTime));
     else if (!strncmp_P(text,PSTR("#MMDD"),5)) sprintf_P(text, zero?PSTR("%02d/%02d")     :PSTR("%d/%d"),      month(localTime), day(localTime));
     else if (!strncmp_P(text,PSTR("#TIME"),5)) sprintf_P(text, zero?PSTR("%02d:%02d%s")   :PSTR("%2d:%02d%s"), AmPmHour,         minute(localTime), sec);
-    else if (!strncmp_P(text,PSTR("#HHMM"),5)) sprintf_P(text, zero?PSTR("%02d:%02d")     :PSTR("%d:%02d"),    AmPmHour,         minute(localTime));
-    else if (!strncmp_P(text,PSTR("#HH"),3))   sprintf_P(text, zero?PSTR("%02d")          :PSTR("%d"),         AmPmHour);
```
Comment (softhack007) [inline wled00/FX.cpp:5964]: Why did you remove "#HHMM` and `#HH` ?
Some user might have them in their preset. We should stay backwards-compatible.

File: wled00/FX.cpp [hunk 2]
```diff
@@ -5960,10 +5960,16 @@ uint16_t mode_2Dscrollingtext(void) {
...
-    else if (!strncmp_P(text,PSTR("#MM"),3))   sprintf_P(text, zero?PSTR("%02d")          :PSTR("%d"),        minute(localTime));
```
Comment (softhack007) [inline wled00/FX.cpp:5965]: Same problem here - changing the managing of `#MM` from "minutes" to "month" is not backwards-compatible either.

File: wled00/FX.cpp [hunk 3]
```diff
@@ -5960,10 +5960,16 @@ uint16_t mode_2Dscrollingtext(void) {
...
-  }
+    else if (!strncmp_P(text,PSTR("#hhmm"),5)) sprintf_P(text, zero?PSTR("%02d:%02d")     :PSTR("%d:%02d"),    AmPmHour,         minute(localTime));
+    else if (!strncmp_P(text,PSTR("#DD"),3))   sprintf_P(text, zero?PSTR("%02d")          :PSTR("%d"),        day(localTime));
+    else if (!strncmp_P(text,PSTR("#MMM"),4))  sprintf_P(text, zero?PSTR("%s")            :PSTR("%s"),        monthShortStr(month(localTime)));
+    else if (!strncmp_P(text,PSTR("#MM"),3))   sprintf_P(text, zero?PSTR("%02d")          :PSTR("%d"),        month(localTime));
+    else if (!strncmp_P(text,PSTR("#YYYY"),5)) sprintf_P(text, zero?PSTR("%04d")          :PSTR("%d"),        year(localTime));
+    else if (!strncmp_P(text,PSTR("#YY"),3))   sprintf_P(text, zero?PSTR("%02d")          :PSTR("%d"),        year(localTime)-2000);
```
Comment (blazoncek) [inline wled00/FX.cpp:5968]: Perhaps a better way is `year(localTime)%100` instead of subtracting 2000.

**Review by blazoncek — CHANGES_REQUESTED** [review]
I would suggest you uppercase `text` and then do formatting.
I.e.
```c++
if (text[0] == '#') for (auto &c : text) c = std::toupper(c);
```

File: wled00/FX.cpp (hunk 3 above)
Comment (5chubrakete) [reply wled00/FX.cpp:5968]: Agreed on %100 way and keeping it backwards-compatible. I dont get the upper case part.
So keep #HH, #MM and add #MON, #MONTH, #SS, #YY, #YEAR ?

**Comment by blazoncek** [comment]
Here's a nonbreaking change:
```c++
    if (text[0] == '#') for (auto &c : text) c = std::toupper(c);
    if      (!strncmp_P(text,PSTR("#DATE"),5)) sprintf_P(text, zero?PSTR("%02d.%02d.%04d"):PSTR("%d.%d.%d"),   day(localTime),   month(localTime),  year(localTime));
    else if (!strncmp_P(text,PSTR("#DDMM"),5)) sprintf_P(text, zero?PSTR("%02d.%02d")     :PSTR("%d.%d"),      day(localTime),   month(localTime));
    else if (!strncmp_P(text,PSTR("#MMDD"),5)) sprintf_P(text, zero?PSTR("%02d/%02d")     :PSTR("%d/%d"),      month(localTime), day(localTime));
    else if (!strncmp_P(text,PSTR("#TIME"),5)) sprintf_P(text, zero?PSTR("%02d:%02d%s")   :PSTR("%2d:%02d%s"), AmPmHour,         minute(localTime), sec);
    else if (!strncmp_P(text,PSTR("#HHMM"),5)) sprintf_P(text, zero?PSTR("%02d:%02d")     :PSTR("%d:%02d"),    AmPmHour,         minute(localTime));
    else if (!strncmp_P(text,PSTR("#HH"),3))   sprintf  (text, zero?    ("%02d")          :    ("%d"),         AmPmHour);
    else if (!strncmp_P(text,PSTR("#MM"),3))   sprintf  (text, zero?    ("%02d")          :    ("%d"),         minute(localTime));
    else if (!strncmp_P(text,PSTR("#SS"),3))   sprintf  (text,          ("%02d")                     ,         second(localTime));
    else if (!strncmp_P(text,PSTR("#DD"),3))   sprintf  (text, zero?    ("%02d")          :    ("%d"),         day(localTime));
    else if (!strncmp_P(text,PSTR("#DAY"),4))  sprintf  (text,          ("%s")                       ,         dayShortStr(day(localTime)));
    else if (!strncmp_P(text,PSTR("#DDDD"),5)) sprintf  (text,          ("%s")                       ,         dayStr(day(localTime)));
    else if (!strncmp_P(text,PSTR("#MO"),3))   sprintf  (text, zero?    ("%02d")          :    ("%d"),         month(localTime));
    else if (!strncmp_P(text,PSTR("#MON"),4))  sprintf  (text,          ("%s")                       ,         monthShortStr(month(localTime)));
    else if (!strncmp_P(text,PSTR("#MMMM"),5)) sprintf  (text,          ("%s")                       ,         monthStr(month(localTime)));
    else if (!strncmp_P(text,PSTR("#YY"),3))   sprintf  (text,          ("%02d")                     ,         year(localTime)%100);
    else if (!strncmp_P(text,PSTR("#YYYY"),5)) sprintf_P(text, zero?PSTR("%04d")          :    ("%d"),         year(localTime));
```

**Review by blazoncek — APPROVED** [review]

**Comment by blazoncek** [comment]
Please rebase this for 0_15 branch.

**Comment by softhack007** [comment]
replaced by #4195.
